server: rework Gossip bootstrap sequence

tbg opened the pull request:
This is a WIP because I want to add the script from #19427, which reproduces
this problem with 100% accuracy, as an acceptance test.

----

A long time ago, in #5441, we changed the server startup sequence so that we
would only open the main listener at the very end of the startup sequence. The
impetus was that by opening the listener early, various services could receive
requests when they weren't actually ready, which resulted in unexpected errors
during tests.

What we missed at the time was that this isn't actually a good idea because it
also robs the node of the ability to accept incoming Gossip connections. As a
result, if a node didn't manage to talk to Gossip peers (perhaps it was started
without a join flag and had nothing reasonable persisted), it would never get
its Gossip connected and would never leave `(*Node).start` which has to wait for
Gossip.

These deadlocks worsened the impact of some of our other insufficiently
streamlined aspects of the startup sequence, namely the choice of `--join`
flags, and were generally hard to debug especially since the HTTP endpoints, and
thus many debugging facilities, weren't available.

This commit reworks the sequence so that

- the HTTP endpoint opens early (so that the debug endpoints are accessible)
- opens the main cmux listener early
    - but delays opening the Postgres sublistener
    - but fails all non-Gossip and non-Ping gRPC traffic retryably

Fixes #18027.
Fixes #13027.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/19493)
<!-- Reviewable:end -->

tbg committed:
server: rework Gossip bootstrap sequence

A long time ago, in #5441, we changed the server startup sequence so that we
would only open the main listener at the very end. The impetus was that by
opening the listener early, various services could receive requests when they
weren't actually ready, which resulted in unexpected errors during tests.

What we missed at the time was that this wasn't actually a good idea because it
also robs the node of the ability to accept incoming Gossip connections. As a
result, if a node didn't manage to talk to Gossip peers (perhaps it was started
without a join flag and had nothing reasonable persisted), it would never
successfully connect to Gossip and would never leave `(*Node).start`.

These deadlocks added to the impact of some of our other insufficiently
streamlined aspects of the startup sequence, namely the choice of `--join`
flags, and were generally hard to debug especially since the HTTP endpoints, and
thus many debugging facilities, weren't available.

This commit reworks the sequence so that

- the HTTP endpoint opens early (so that the debug endpoints are accessible)
- the main cmux listener opens early
    - but opening the Postgres sublistener is delayed
    - and all non-Gossip and non-Ping gRPC traffic fails retryably.

@bdarnell points out:

    Some more important context is that in 1.0, the server didn't open its port
    until it was ready to serve, so TCP-level health checks were sufficient. With
    the introduction of the init command in 1.1, a not-yet-initialized server will
    open its port early (which means that higher-level health checks are required
    for determining readiness). Now that we do that, we might as well also have
    initialized-but-not-connected servers do the same.

Fixes #18027.
Fixes #13027.
Fixes #19427.

bdarnell commented:
LGTM (I haven't read line-by-line but I assume it's basically just code movement)

Some more important context is that in 1.0, the server didn't open its port until it was ready to serve, so TCP-level health checks were sufficient. With the introduction of the `init` command in 1.1, a not-yet-initialized server will open its port early (which means that higher-level health checks are required for determining readiness). Now that we do that, we might as well also have initialized-but-not-connected servers do the same. 

---

Review status: 0 of 2 files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---

*[pkg/server/server.go, line 200 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/19493#-KxEUJk9mjpkBT0KkHXu:-KxEUJk9mjpkBT0KkHXv:bdm1u12) ([raw file](https://github.com/cockroachdb/cockroach/blob/6e533905830421c082ce6b839a9f96a374a686ef/pkg/server/server.go#L200)):*
> ```Go
> 
> 	{
> 		var earlyRPCWhiteList = map[string]struct{}{
> ```

Can we unify this with the init server? It would be nice to only have one version of this interceptor. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/19493#-:-KxETA_7wLO5oS9GWDxs:b-fyu2xs)*
<!-- Sent from Reviewable.io -->

a-robinson commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>, but ditto on having glossed over some of the code movement. Have you verified that this fixes situations like your gist in #19427?

---

Reviewed 2 of 2 files at r1.
Review status: all files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

*[pkg/server/server.go, line 1103 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/19493#-KxEqE8-02H9oX3qlvJe:-KxEqE8-02H9oX3qlvJf:b-r0v0u8) ([raw file](https://github.com/cockroachdb/cockroach/blob/26030c8a608ab8c8b5969840aef896e24d4aec80/pkg/server/server.go#L1103)):*
> 19427

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/19493#-:-KxEqSkoCpR_SXeJo147:b-chszww)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
> Have you verified that this fixes situations like your gist in #19427?

Yep, by now I finished translating that into an acceptance test, too.

> Some more important context is that in 1.0, the server didn't open its port until it was ready to serve, so TCP-level health checks were sufficient. With the introduction of the init command in 1.1, a not-yet-initialized server will open its port early (which means that higher-level health checks are required for determining readiness). Now that we do that, we might as well also have initialized-but-not-connected servers do the same.

Anything actionable in this PR? Should the health endpoint be opened later?

PTAL at the new test.

---

Review status: 1 of 7 files reviewed at latest revision, 2 unresolved discussions, some commit checks pending.

---

*[pkg/server/server.go, line 200 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/19493#-KxEUJk9mjpkBT0KkHXu:-KxJ-8cyMxp1RrqtRUQ-:bey20oo) ([raw file](https://github.com/cockroachdb/cockroach/blob/6e533905830421c082ce6b839a9f96a374a686ef/pkg/server/server.go#L200)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

Can we unify this with the init server? It would be nice to only have one version of this interceptor. 
</blockquote></details>

Good idea, though I'll file this for now: https://github.com/cockroachdb/cockroach/issues/19523

---

*[pkg/server/server.go, line 1103 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/19493#-KxEqE8-02H9oX3qlvJe:-KxIqmqu6B1i-yboZuSf:b-4wrawl) ([raw file](https://github.com/cockroachdb/cockroach/blob/26030c8a608ab8c8b5969840aef896e24d4aec80/pkg/server/server.go#L1103)):*
<details><summary><i>Previously, a-robinson (Alex Robinson) wrote…</i></summary><blockquote>

> 19427
</blockquote></details>

Happy little accident?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/19493#-:-KxIpDa8a_HCDoVtU2m8:b-pwalof)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
No, I wasn't suggesting any changes to this PR, but just thought that comment might be worth preserving somewhere (maybe in the commit message?)

---

Reviewed 1 of 1 files at r2, 4 of 5 files at r3, 1 of 1 files at r4.
Review status: all files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---

*[pkg/acceptance/gossip_peerings_test.go, line 189 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/19493#-KxJ56yfX6Vi3vLVhOLM:-KxJ56yfX6Vi3vLVhOLN:b-15caap) ([raw file](https://github.com/cockroachdb/cockroach/blob/d70cd316867853589727289ded96650177bc2781/pkg/acceptance/gossip_peerings_test.go#L189)):*
> ```Go
> 
> // This test verifies that when the first node is restarted and has no valid
> // join flags or persisted Gossip bootstrap records, it will still be able to
> ```

Why doesn't the first node have persisted gossip bootstrap records?

---

*[pkg/acceptance/gossip_peerings_test.go, line 217 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/19493#-KxJ4m76QVbCHBEH-uUi:-KxJ4m76QVbCHBEH-uUj:b-qcuu2k) ([raw file](https://github.com/cockroachdb/cockroach/blob/d70cd316867853589727289ded96650177bc2781/pkg/acceptance/gossip_peerings_test.go#L217)):*
> ```Go
> 
> 	lc := c.(*localcluster.LocalCluster)
> 	// Upgrade the first node's binary to match the other nodes (i.e. the testing binary).
> ```

Copy-pasta comment?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/19493#-:-KxJ3mcWKiRbAxvui__F:b-bb99e1)*
<!-- Sent from Reviewable.io -->

tbg committed:
acceptance: verify isolated first node accepts incoming Gossip

See #18027.

tschottdorf commented:
Ack, done. TFTRs!

---

Review status: all files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---

*[pkg/acceptance/gossip_peerings_test.go, line 189 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/19493#-KxJ56yfX6Vi3vLVhOLM:-KxJ7QtFOO_YcDoJSw-F:blm6a8h) ([raw file](https://github.com/cockroachdb/cockroach/blob/d70cd316867853589727289ded96650177bc2781/pkg/acceptance/gossip_peerings_test.go#L189)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

Why doesn't the first node have persisted gossip bootstrap records?
</blockquote></details>

The test sets up things so that the first node's bootstrap info is all outdated. Added comment.

---

*[pkg/acceptance/gossip_peerings_test.go, line 217 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/19493#-KxJ4m76QVbCHBEH-uUi:-KxJ7a3DsAxatAdE9mYS:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d70cd316867853589727289ded96650177bc2781/pkg/acceptance/gossip_peerings_test.go#L217)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

Copy-pasta comment?
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/19493#-:-KxJ8Tcje2IbPmIGwOza:b-swe23a)*
<!-- Sent from Reviewable.io -->

a-robinson commented:
Review status: 2 of 7 files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---

*[pkg/server/server.go, line 1103 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/19493#-KxEqE8-02H9oX3qlvJe:-KxJ7T_92KiXHBIkLtyZ:b-rgzsmz) ([raw file](https://github.com/cockroachdb/cockroach/blob/26030c8a608ab8c8b5969840aef896e24d4aec80/pkg/server/server.go#L1103)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

Happy little accident?
</blockquote></details>

"Happy little accident" could be a good replacement for the shame cube

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/19493)*
<!-- Sent from Reviewable.io -->

tamird commented on pkg/rpc/context.go:
```diff
@@ -131,15 +140,64 @@ func NewServer(ctx *Context) *grpc.Server {
+	// functions a wee bit more performant by hoisting some of the nil checks
+	// out. Doubt measurements can tell the difference though.
+
+	if interceptor != nil {
```
Some of this code might benefit from https://godoc.org/github.com/grpc-ecosystem/go-grpc-middleware ?

tschottdorf commented:
@bdarnell what are your thoughts on cherry-picking this to 1.1.2? I think it can wait, though the issue that is being fixed is sure to cause a lot of problems when it occurs.

bdarnell commented:
I'm reluctant to make this change in a patch release; this is a potentially disruptive change (see #19559. It could also require changes to LB configurations). However, this depends on the availability of other solutions, since the problem being solved is pretty severe. If you follow best practices in setting join flags, how likely are you to run into this problem? Does it only occur if there are IP changes, or could it be more common?

a-robinson commented:
To me, this seems too big/risky of a change to be patched into 1.1.

As far as I'm aware, the only situations we've constructed in which it becomes a problem are when IPs or ports change, which shouldn't be common in real deployments.

tschottdorf commented:
I'm ok letting this bake until 2.0. I don't think you run into this much outside of the intentionally exotic things our acceptance tests do.
